### PR TITLE
Ntt 162 - Beneficiaries frontend

### PR DESCRIPTION
**Pull request by matthewharris1**
This should only include files changed after Nunjucks conversion took place as that was merged before hand. 

Something to point out. I have made the Index controller view etc. into InterruptInfo as it behaves the exact same way and is the first page in the flow currently. 

Test coverage above 80% as seen below.

Screenshot 2020-07-15 at 13 34 26

### CODE REVIEW

File: app/models/TypeOfBeneficiary.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  case object Unidentified extends WithName("unidentified") with TypeOfBeneficiary
+  case object Charity extends WithName("charity") with TypeOfBeneficiary
+  case object Trust extends WithName("trust") with TypeOfBeneficiary
+  case object Company extends WithName("company") with TypeOfBeneficiary
+  case object LargeNumbers extends WithName("largeNumbers") with TypeOfBeneficiary
+  case object Other extends WithName("other") with TypeOfBeneficiary
+
+  val values: Seq[TypeOfBeneficiary] = Seq(
+    Individual,
+    Unidentified
```
Comment (parkeradam) [inline app/models/TypeOfBeneficiary.scala:38]: Youve missed adding the rest of the options to the list here

File: app/utils/CheckYourAnswersHelper.scala [hunk 2]
```diff
@@ -23,11 +23,315 @@ import models.{CheckMode, UserAnswers}
...
 import CheckYourAnswersHelper._
+import services.CountryService
 import uk.gov.hmrc.viewmodels._
 import uk.gov.hmrc.viewmodels.SummaryList._
 import uk.gov.hmrc.viewmodels.Text.Literal
 
-class CheckYourAnswersHelper(userAnswers: UserAnswers)(implicit messages: Messages) {
+class CheckYourAnswersHelper(userAnswers: UserAnswers, countryService: CountryService)(implicit messages: Messages) {
+
+  private def country(code: String): Content =
```
Comment (parkeradam) [inline app/utils/CheckYourAnswersHelper.scala:33]: private functions should be below public ones

**Comment by parkeradam** [comment]
More general comment, I'd rather the "UnidentifiedDescription" and "BigNumberName" have the word class in them (like they do the ticket) as it confused me reading through not knowing they were classifications. Helps to further direct what they actually are.

File: conf/app.routes [hunk 3]
```diff
@@ -13,3 +11,107 @@ GET        /check-your-answers                          controllers.CheckYourAns
 GET        /unauthorised                                controllers.UnauthorisedController.onPageLoad
 
 GET        /accessibility-statement                     controllers.AccessibilityController.onPageLoad
+
+GET        /typeOfBeneficiary                        controllers.TypeOfBeneficiaryController.onPageLoad(mode: Mode = NormalMode)
+POST       /typeOfBeneficiary                        controllers.TypeOfBeneficiaryController.onSubmit(mode: Mode = NormalMode)
+GET        /changeTypeOfBeneficiary                  controllers.TypeOfBeneficiaryController.onPageLoad(mode: Mode = CheckMode)
+POST       /changeTypeOfBeneficiary                  controllers.TypeOfBeneficiaryController.onSubmit(mode: Mode = CheckMode)
+
+GET        /dateOfBirth                  controllers.DateOfBirthController.onPageLoad(mode: Mode = NormalMode)
```
Comment (owen-opencast) [inline conf/app.routes:20]: align the right-hand column of routes

Comment (matthewharris1) [reply conf/app.routes]: Done

File: app/assets/stylesheets/application.scss [hunk 4]
```diff
@@ -2,3 +2,7 @@ $govuk-assets-path: "/ntt-beneficiaries-frontend/assets/lib/govuk-frontend/govuk
 
 @import "lib/govuk-frontend/govuk/all";
 @import "language-toggles.scss";
+
+ul#country-autocomplete__listbox {
+  font-family: arial;
+}
```
Comment (ajchilts) [inline app/assets/stylesheets/application.scss:8]: ...and here

File: app/controllers/CheckYourAnswersController.scala [hunk 5]
```diff
@@ -22,31 +22,68 @@ import play.api.i18n.{I18nSupport, MessagesApi}
...
         "check-your-answers.njk",
         Json.obj("list" -> answers)
       ).map(Ok(_))
   }
+
+  def onSubmit: Action[AnyContent] = (identify andThen getData andThen requireData).async {
+    implicit request =>
+      val answers = UserAnswers(request.internalId)
+      for {
+        _              <- sessionRepository.set(answers)
```
Comment (ajchilts) [inline app/controllers/CheckYourAnswersController.scala:86]: Please remove whitespace

File: app/controllers/CountryLookup.scala [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  private def getMsg(messages: Messages, defaultMessage:Option[String]):String =
+    defaultMessage.map(msg => messages(msg)).getOrElse("")
+
+  def countries(messages: Messages, selected: Option[String] = None, defaultMessage:Option[String] = None): Seq[JsObject] =
+    Seq(Json.obj("value" -> "", "text" -> getMsg(messages, defaultMessage))) ++
+      countryService.getCountries.getOrElse(Seq.empty).map(nv => {
+        val json = Json.obj(("value", nv.value), ("text", nv.name))
+        selected.map(s => if(nv.value == s) json + selectedField else json).getOrElse(json)
+      })
+}
```
Comment (ajchilts) [inline app/controllers/CountryLookup.scala:37]: Please add a newline at the end

File: app/models/Description.scala [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    import play.api.libs.functional.syntax._
+    (
+      (__ \ "descriptionOne").write[String] and
+        (__ \ "descriptionTwo").writeNullable[String] and
+        (__ \ "descriptionThree").writeNullable[String] and
+        (__ \ "descriptionFour").writeNullable[String] and
+        (__ \ "descriptionFive").writeNullable[String]
+      ) (unlift(Description.unapply))
+  }
+}
```
Comment (ajchilts) [inline app/models/Description.scala:50]: Newline

File: app/models/Name.scala [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+  implicit lazy val writes: OWrites[Name] = {
+    import play.api.libs.functional.syntax._
+    (
+      (__ \ "firstName").write[String] and
+        (__ \ "middleName").writeNullable[String] and
+        (__ \ "lastName").write[String]
+      ) (unlift(Name.unapply))
+  }
+}
```
Comment (ajchilts) [inline app/models/Name.scala:44]: Same here

File: app/services/CountryService.scala [hunk 9]
```diff
@@ -14,24 +14,17 @@
...
 
-  def onPageLoad: Action[AnyContent] = Action.async {
-    implicit request =>
+  def getCountryByCode(code: String): Option[String] = getCountries
+    .flatMap(seq => seq.find(nv => nv.value == code).map(nv => nv.name))
 
-      renderer.render("index.njk").map(Ok(_))
-  }
-}
+}
```
Comment (ajchilts) [inline app/services/CountryService.scala:30]: Newline

File: conf/app.routes [hunk 10]
```diff
@@ -1,15 +1,123 @@
...
+
+GET        /knownCountry                                controllers.KnownCountryController.onPageLoad(mode: Mode = NormalMode)
+POST       /knownCountry                                controllers.KnownCountryController.onSubmit(mode: Mode = NormalMode)
+GET        /changeKnownCountry                          controllers.KnownCountryController.onPageLoad(mode: Mode = CheckMode)
+POST       /changeKnownCountry                          controllers.KnownCountryController.onSubmit(mode: Mode = CheckMode)
+
+GET        /interruptInfo                               controllers.InterruptInfoController.onPageLoad(mode: Mode = NormalMode)
+POST       /interruptInfo                               controllers.InterruptInfoController.onSubmit(mode: Mode = NormalMode)
+
+GET        /addMore                        controllers.AddMoreController.onPageLoad(mode: Mode = NormalMode)
```
Comment (ajchilts) [inline conf/app.routes:120]: Please correct indentation here

File: conf/views/interruptInfo.njk [hunk 11]
```diff
@@ -13,11 +13,20 @@
...
+            }) }}
+
+        </form>
+
       </div>
     </div>
   </div>
-
-{% endblock %}
+{% endblock %}
```
Comment (ajchilts) [inline conf/views/interruptInfo.njk:32]: Newline

File: app/assets/data/countries/des-codes.json [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+  "CC","CD","CF","CG","CH","CI","CK","CL","CM","CN","CO","CR","CS","CU","CV","CW","CX","CY","CZ","DE","DJ","DK","DM",
+  "DO","DZ","EC","EE","EG","EH","ER","ES","ET","FC","FI","FJ","FK","FM","FO","FR","GA","GB","GD","GE","GF","GG","GH",
+  "GI","GL","GM","GN","GP","GQ","GR","GS","GT","GU","GW","GY","HK","HM","HN","HR","HT","HU","ID","IE","IL","IM","IN",
+  "IO","IQ","IR","IS","IT","JE","JM","JO","JP","KE","KG","KH","KI","KM","KN","KP","KR","KW","KY","KZ","LA","LB","LC",
+  "LI","LK","LR","LS","LT","LU","LV","LY","MA","MC","MD","ME","MF","MG","MH","MK","ML","MM","MN","MO","MP","MQ","MR",
+  "MS","MT","MU","MV","MW","MX","MY","MZ","NA","NC","NE","NF","NG","NI","NL","NO","NP","NR","NU","NZ","OM","PA","PE",
+  "PF","PG","PH","PK","PL","PM","PN","PR","PS","PT","PW","PY","QA","RE","RO","RS","RU","RW","SA","SB","SC","SD","SE",
+  "SG","SH","SI","SJ","SK","SL","SM","SN","SO","SR","SS","ST","SV","SX","SY","SZ","TC","TD","TF","TG","TH","TJ","TK",
+  "TL","TM","TN","TO","TP","TR","TT","TV","TW","TZ","UA","UG","UM","US","UY","UZ","VA","VC","VE","VG","VI","VN","VU",
+  "WF","WS","YE","YT","ZA","ZM","ZW"]
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/assets/data/countries/des-codes.json]: Done

File: app/assets/js/countries-autocomplete.js [hunk 13]
```diff
@@ -0,0 +1,3 @@
+accessibleAutocomplete.enhanceSelectElement({
+    selectElement: document.querySelector('#country-autocomplete')
+})
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/assets/js/countries-autocomplete.js]: Done

File: app/assets/stylesheets/application.scss (hunk 4 above)
Comment (matthewharris1) [reply app/assets/stylesheets/application.scss]: Done

File: app/controllers/CheckYourAnswersController.scala (hunk 5 above)
Comment (matthewharris1) [reply app/controllers/CheckYourAnswersController.scala]: Done

File: app/controllers/CountryLookup.scala (hunk 6 above)
Comment (matthewharris1) [reply app/controllers/CountryLookup.scala]: Done

File: app/models/Description.scala (hunk 7 above)
Comment (matthewharris1) [reply app/models/Description.scala]: Done

File: app/models/Name.scala (hunk 8 above)
Comment (matthewharris1) [reply app/models/Name.scala]: Done

File: app/models/autocomplete/CountryDataProvider.scala [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+
+  override def fetch: Option[Seq[NameValuePair]] = json map { j =>
+    j.as[Seq[NameValuePair]] map {
+      case n@NameValuePair(_, value) if isEncoded(value) => n.copy(value = stripCode(value))
+      case n => n
+    } collect {
+      case n if countryCodes.contains(n.value.toUpperCase()) => n
+    }
+  }
+}
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/models/autocomplete/CountryDataProvider.scala]: Done

File: app/models/autocomplete/NameValuePair.scala [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+  implicit val jsonReads: Reads[NameValuePair] = {
+    __.read[Seq[String]] map {
+      case key :: value :: _ => NameValuePair(key, value)
+    }
+  }
+
+  implicit val jsonWrites = Writes[NameValuePair] { model =>
+    JsArray(Seq(JsString(model.name), JsString(model.value)))
+  }
+}
```
[the remark this answers is not in the document]
Comment (matthewharris1) [reply app/models/autocomplete/NameValuePair.scala]: Done

File: app/services/CountryService.scala (hunk 9 above)
Comment (matthewharris1) [reply app/services/CountryService.scala]: Done

File: conf/app.routes (hunk 10 above)
Comment (matthewharris1) [reply conf/app.routes]: Done

File: conf/views/interruptInfo.njk (hunk 11 above)
Comment (matthewharris1) [reply conf/views/interruptInfo.njk]: Done

File: app/utils/CheckYourAnswersHelper.scala [hunk 16]
```diff
@@ -19,15 +19,337 @@ package utils
...
+          )
+        )
+      )
+  }
+
+  def description: Option[Row] = userAnswers.get(DescriptionPage) map {
+    answer =>
+      Row(
+        key     = Key(msg"description.checkYourAnswersLabel", classes = Seq("govuk-!-width-one-half")),
+        value   = Value(lit"$answer"),
```
Comment (owen-opencast) [inline app/utils/CheckYourAnswersHelper.scala:97]: Description is displaying incorrectly

File: app/forms/AddMoreFormProvider.scala [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+package forms
+
+import javax.inject.Inject
+
+import forms.mappings.Mappings
+import play.api.data.Form
+
+class AddMoreFormProvider @Inject() extends Mappings {
+
+  def apply(): Form[String] =
```
Comment (owen-opencast) [inline app/forms/AddMoreFormProvider.scala:26]: Change to a radio button page and add the buttons from the screenshot

Comment (matthewharris1) [reply app/forms/AddMoreFormProvider.scala]: Done

File: app/utils/CheckYourAnswersHelper.scala (hunk 16 above)
Comment (matthewharris1) [reply app/utils/CheckYourAnswersHelper.scala]: Done

File: app/utils/CheckYourAnswersHelper.scala [hunk 18]
```diff
@@ -19,22 +19,375 @@ package utils
...
   private def yesOrNo(answer: Boolean): Content =
     if (answer) {
       msg"site.yes"
     } else {
       msg"site.no"
     }
+
+  private def descriptionHtml(x: List[Option[String]]): Html = {
+
+    val ifBullet = if (flatList.size == 1) "<ul class=\"list\">" else "<ul class=\"list list-bullet\">"
```
Comment (owen-opencast) [inline app/utils/CheckYourAnswersHelper.scala:363]: Compilation error here - forward reference to flatList

**Pull request merged**